cql3/maps.cc

```diff
@@ -158,15 +158,13 @@ maps::value::from_serialized(const fragmented_temporary_buffer::view& fragmented
         // Collections have this small hack that validate cannot be called on a serialized object,
         // but compose does the validation (so we're fine).
         // FIXME: deserialize_for_native_protocol?!
-      return with_linearized(fragmented_value, [&] (bytes_view value) {
-        auto m = value_cast<map_type_impl::native_type>(type.deserialize(value, sf));
+        auto m = value_cast<map_type_impl::native_type>(type.deserialize(fragmented_value, sf));
```

Comment:
In many cases, the value will come in a single fragment (e.g. someone passed us an int32, the odds of it being fragmented in our 8k buffers are 1:2000). But we will pay the cost of fragmented deserialization nevertheless (endless checking).

So abstract_type::deserialize() can check whether we have 0 or 1 fragments, and call the bytes_view variant in those cases.

Or maybe it can be a helper function that takes in a fragmented view and a function, checks if it has more than one fragment, and calls the inner function with a single_fragment_view if possible, and with the original input if not. This may replace many checks during inner processing with just one check in the outer processing.

This can come post merge.

Reply:
I think long-term we should examine why do we need these serialize-deserialize methods at all. We need to be able to do 3 things with our values:
* de/serialize to CQL wire format: for most types this is the same as the in-memory format
* de/serialize to on-disk format: I suspect for most types the same applies as before
* compare types

Neither of these should require deserializing a fragmented buffer belonging to a UTF8 value into a `sstring` C++ type, then re-serializing it into another format or comparing it to another `sstring`. We should drop the intermediate format.

One more uses case is pretty printing and genuine conversion to C++ types in tests. Neither of these is on any hot path.

Reply:
@avikivity Done in c990869. Please take a look.